rust/zypp-agama/src/lib.rs

```diff
@@ -239,6 +239,75 @@ pub fn unselect_resolvable(name: &str, kind: ResolvableKind) -> ZyppResult<()> {
     }
 }
 
+#[derive(Debug)]
+pub enum ResolvableSelected {
+    NotSelected,
+    UserSelected,
+    InstallationSelected,
+    SolverSelected,
+}
+
+impl From<zypp_agama_sys::RESOLVABLE_SELECTED> for ResolvableSelected {
+    fn from(value: zypp_agama_sys::RESOLVABLE_SELECTED) -> Self {
+        match value {
+            zypp_agama_sys::RESOLVABLE_SELECTED_NOT_SELECTED => Self::NotSelected,
+            zypp_agama_sys::RESOLVABLE_SELECTED_USER_SELECTED => Self::UserSelected,
+            zypp_agama_sys::RESOLVABLE_SELECTED_INSTALLATION_SELECTED => Self::InstallationSelected,
+            zypp_agama_sys::RESOLVABLE_SELECTED_SOLVER_SELECTED => Self::SolverSelected,
+            _ => Self::SolverSelected, // XXX: should not happen
+        }
+    }
+}
+
+// TODO: should we add also e.g. serd serializers here?
+#[derive(Debug)]
+pub struct PatternInfo {
+    pub name: String,
+    pub category: String,
+    pub icon: String,
+    pub description: String,
+    pub summary: String,
+    pub order: String,
+    pub selected: ResolvableSelected,
+}
+
+pub fn patterns_info(names: Vec<&str>) -> ZyppResult<Vec<PatternInfo>> {
+    unsafe {
+        let mut status: Status = Status::default();
+        let status_ptr = &mut status as *mut _;
+        let c_names: Vec<CString> = names.iter()
+            .map(|s| CString::new(*s).expect("CString failed")).collect();
```

Comment:
NP, I just learned how it can fail so we can improve the message a bit
```suggestion
            .map(|s| CString::new(*s).expect("CString must not contain internal NUL")).collect();
```